modules/workflow-service-impl/src/main/java/org/opencastproject/workflow/impl/WorkflowServiceImpl.java

```diff
@@ -357,12 +359,23 @@ protected void fireListeners(final WorkflowInstance oldWorkflowInstance, final W
    */
   @Override
   public List<WorkflowDefinition> listAvailableWorkflowDefinitions() {
-    List<WorkflowDefinition> list = new ArrayList<>();
-    for (Entry<String, WorkflowDefinition> entry : workflowDefinitionScanner.getWorkflowDefinitions().entrySet()) {
-      list.add(entry.getValue());
-    }
-    Collections.sort(list); //sorts by title
-    return list;
+    final String organization = securityService.getOrganization().getId();
+    return workflowDefinitionScanner.getWorkflowDefinitionsStream().filter(x ->
+            x.getOrganization() == null && !workflowDefinitionScanner
+                    .containsIdentifier(new WorkflowIdentifier(x.getId(), organization)) || organization
+                    .equals(x.getOrganization()))
+            .map(WorkflowDefinition::getId)
+            .map(x -> {
+              final WorkflowDefinition definition = workflowDefinitionScanner
+                      .getWorkflowDefinition(new WorkflowIdentifier(x, organization));
+              if (definition != null) {
+                return definition;
+              }
+              return Objects.requireNonNull(
+                      workflowDefinitionScanner.getWorkflowDefinition(WorkflowIdentifier.withoutOrganization(x)));
+            })
+            .sorted()
+            .collect(Collectors.toList());
```

Comment:
This is quite complicated. So we first build workflowIdentifiers based on the workflowDefinitions and then ask the workflowDefinitionScanner for each WorkflowDefinition seperately and sometimes twice? Maybe we should integrate this functionality (getWorkflowDefinitions(Organization org)) into the workflowDefinitionScanner to make it simpler.

Reply:
While the complexity is somewhat inherent because of the fallback mechanism, I was able to come up with a better solution.